Fixes "CSR parsing with critical fields fails"

mschulz-at-hilscher opened the pull request:
## Description

Fixes #8377. Parsing CSRs with extensions that are optionally marked as critical does not fail anymore.

## PR checklist

- [X] **changelog** is provided
- [x] **backport** see: https://github.com/Mbed-TLS/mbedtls/pull/8491
- [X] **tests** are provided

tom-cosgrove-arm commented on library/x509_csr.c:
```diff
@@ -78,6 +78,7 @@ static int x509_csr_parse_extensions(mbedtls_x509_csr *csr,
     int ret;
     size_t len;
     unsigned char *end_ext_data;
+    int critical;
```
(code style) please leave a blank line between the declared variables and the body of the function

mschulz-at-hilscher replied:
Done

tom-cosgrove-arm commented:
This really looks like something that should have a test

mschulz-at-hilscher commented:
> This really looks like something that should have a test

Shall I create a new CSR test file under data_files and add a corresponding test case to

paul-elliott-arm commented on library/x509_csr.c:
```diff
@@ -100,6 +102,9 @@ static int x509_csr_parse_extensions(mbedtls_x509_csr *csr,
         *p += extn_oid.len;
 
+        /* Get and ignore optional critical flag */
+        (void)mbedtls_asn1_get_bool(p, end_ext_data, &critical);
```
I'm not happy with us just ignoring critical extensions that we don't support as I don't think that is correct for the spec. This wouldn't be too difficult to fix - stash the value of critical here (but default to false), then in the switch just down from here where we would normally skip an unsupported extension, do something like:
```
return MBEDTLS_ERROR_ADD(MBEDTLS_ERR_X509_INVALID_EXTENSIONS,
                                         MBEDTLS_ERR_ASN1_INVALID_DATA);
```
I'm not sure if this is the correct error here, but you hopefully get the idea.

mschulz-at-hilscher replied:
Handling is now done similarly to x509_get_crt_ext.

paul-elliott-arm commented on library/x509_csr.c:
```diff
@@ -100,6 +102,9 @@ static int x509_csr_parse_extensions(mbedtls_x509_csr *csr,
         *p += extn_oid.len;
 
+        /* Get and ignore optional critical flag */
+        (void)mbedtls_asn1_get_bool(p, end_ext_data, &critical);
```
```suggestion
        (void) mbedtls_asn1_get_bool(p, end_ext_data, &critical);
```
Code style checker failed as well.

mschulz-at-hilscher replied:
changed.

mschulz-at-hilscher commented:
I think for my usecase I will need a callback function to handle unknown critical extensions.

paul-elliott-arm commented:
The new test you added appears to be failing with the latest code as well - given the test suites are executables, they are easy to debug.

mschulz-at-hilscher commented:
> The new test you added appears to be failing with the latest code as well - given the test suites are executables, they are easy to debug.

The CSRs required on our side contain extensions unsupported by mbedtls. Hence, I now added a callback to handle those extensions. The approach is similar to handling unsupported extensions during certificate parsing. I also added new test cases. Maybe the changelog might need an update as well to reflect that new functions exist to parse CSRs.

paul-elliott-arm commented:
> Maybe the changelog might need an update as well to reflect that new functions exist to parse CSRs.

Yep, definitely as this would certainly be something the end user (you!) would be interested in. I will re-run the tests.

mschulz-at-hilscher commented:
@paul-elliott-arm Do you know why my helper functions were not found during the build? When building locally, it works:

```
suites/test_suite_x509parse.function: In function 'test_mbedtls_x509_csr_parse_with_ext_cb':

[2023-10-18T22:22:48.441Z] suites/test_suite_x509parse.function:1288:62: error: 'parse_csr_ext_accept_cb' undeclared (first use in this function)

[2023-10-18T22:22:48.441Z]                                                      accept ? parse_csr_ext_accept_cb :

[2023-10-18T22:22:48.441Z]                                                               ^

[2023-10-18T22:22:48.441Z] suites/test_suite_x509parse.function:1288:62: note: each undeclared identifier is reported only once for each function it appears in

[2023-10-18T22:22:48.441Z] suites/test_suite_x509parse.function:1289:62: error: 'parse_csr_ext_reject_cb' undeclared (first use in this function)

[2023-10-18T22:22:48.441Z]                                                               parse_csr_ext_reject_cb,

[2023-10-18T22:22:48.441Z]                                                               ^

[2023-10-18T22:22:48.441Z] Makefile:228: recipe for target 'test_suite_x509parse' failed
```

The functions are available in the test function header part: https://github.com/Mbed-TLS/mbedtls/pull/8378/files#diff-fdcf8ae37b642bc33af08b33f6dbe62c1592bc976df8a276af76c030077d0bb3R416

paul-elliott-arm commented:
> Should be solved now. Question just for reference:

Is the final commit the solution for this? Otherwise do you know which particular build this fails on?

paul-elliott-arm commented:
> > Should be solved now. Question just for reference:
> 
> Is the final commit the solution for this? Otherwise do you know which particular build this fails on?

I manually re-ran the previously failing test and it now passes, so hopefully CI will be good this time.

mschulz-at-hilscher commented:
> > Should be solved now. Question just for reference:
> 
> Is the final commit the solution for this? Otherwise do you know which particular build this fails on?

At least, I expect it to be the solution. Sorry for the inconveniences.

paul-elliott-arm commented:
> > > Should be solved now. Question just for reference:
> > 
> > 
> > Is the final commit the solution for this? Otherwise do you know which particular build this fails on?
> 
> At least, I expect it to be the solution. Sorry for the inconveniences.

No inconvenience, don't worry! I was just trying to check whether or not I still needed to test things :)

paul-elliott-arm commented:
Really sorry, the code style has once again complained. If you have uncrustify installed, you can run `scripts/code_style.py `and it will show you the issues. If you pass ` -f ` it will do the fixes for you. Failing that I can either make them as suggestions or perhaps add a commit, let me know.

paul-elliott-arm commented:
Given this is still blocked, I would recommend adding a changelog entry at this point as well - see https://github.com/Mbed-TLS/mbedtls/blob/development/ChangeLog.d/00README.md (you can use `scripts/assemble_changelog.py -o TestChangelog` or something like that to check formatting prior to CI)

mschulz-at-hilscher commented:
> Really sorry, the code style has once again complained. If you have uncrustify installed, you can run `scripts/code_style.py `and it will show you the issues. If you pass `-f` it will do the fixes for you. Failing that I can either make them as suggestions or perhaps add a commit, let me know.

I ran it on the files that I changed.

paul-elliott-arm reviewed (approved):
This looks good to me, thanks for your work on this.

mschulz-at-hilscher commented:
Is there already a plan when this PR will be merged and in which release it will be integrated?

paul-elliott-arm commented:
> Is there already a plan when this PR will be merged and in which release it will be integrated?

I will chase second review again. I am presuming we can get this in for the next release, as we have time for that.

mschulz-at-hilscher commented:
@davidhorstmann-arm, @tom-cosgrove-arm: Any update on when the reviews will be finished and the PR could be merged?

davidhorstmann-arm commented:
@mschulz-at-hilscher I'll provide the other review, but I have a couple of other things on my queue as well.

davidhorstmann-arm reviewed (dismissed):
LGTM

This proved easier to review than expected, so thank you!

gilles-peskine-arm commented:
Why is a backport not required? Does 2.28 not have the bug? Even if it doesn't have the bug, we should at least backport the non-regression test.

Adding "needs-backport" unless there's a justification for not needing to backport the test.

mschulz-at-hilscher commented:
> Why is a backport not required? Does 2.28 not have the bug? Even if it doesn't have the bug, we should at least backport the non-regression test.
> 
> Adding "needs-backport" unless there's a justification for not needing to backport the test.

Hi Gilles,
how does the backporting work? Shall I create a second PR for another branch? Shall it only contain the test cases?
Matthias

gilles-peskine-arm commented:
`mbedtls-2.28` is a [long-time support branch](https://github.com/Mbed-TLS/mbedtls/blob/development/BRANCHES.md#long-time-support-branches), where we fix bugs but we don't add new features.

Normally, if there's a bug fix in `development` for a feature that also exists on the LTS branch, we backport the non-regression test, and we also fix the bug if it's present.

As far as I can tell from a quick look, this is a fix for a bug in CSR parsing, and CSR parsing exists in 2.28. So please make a pull request for `mbedtls-2.28` with at least the new tests, and with the bug fix if necessary.

paul-elliott-arm commented:
Apologies, I was thinking of this too much as a new feature, not an actual bug fix - make a new PR against mbedtls-2.28 with as much of this as you think fit - at minimum we need to fix the bug, and add non-regression tests - however it might be simpler just to backport all of it. Please add me as a reviewer on the new PR, and ping me if you need any more help.

mschulz-at-hilscher commented:
I just checked and realized again why I came up with that PR in the first place. In mbedTLS2 there was no check which extensions were included in a CSR. It was only added in mbedTLS3. That is the reason why we ran into problems when upgrading from mbedTLS2 to mbedTLS3. Hence, I think it does not make sense to backport the changes or any introduced tests to mbedTLS2 as it does not contain the problem.

You can compare the differences here: https://github.com/Mbed-TLS/mbedtls/compare/mbedtls-2.28...development#diff-dccd5ffd55ff529be4e531be08342be60a967c90fa9f1ad0aef790e615846148R347

The x509_csr_parse_attributes function did not exist in mbedTLS2 and there were no other means implemented to perform the checks done in that function in mbedTLS2.

gilles-peskine-arm commented on tests/suites/test_suite_x509parse.data:
```diff
@@ -2940,6 +2940,18 @@ X509 CSR ASN.1 (OK)
 
+X509 CSR ASN.1 (Unsupported critical extension)
+depends_on:MBEDTLS_PK_CAN_ECDSA_SOME:MBEDTLS_ECP_HAVE_SECP256R1:MBEDTLS_MD_CAN_SHA256:!MBEDTLS_X509_REMOVE_INFO
+mbedtls_x509_csr_parse:"308201233081cb02010030413119301706035504030c1053656c66207369676e65642074657374310b300906035504061302444531173015060355040a0c0e41757468437274444220546573743059301306072a8648ce3d020106082a8648ce3d03010703420004c11ebb9951848a436ca2c8a73382f24bbb6c28a92e401d4889b0c361f377b92a8b0497ff2f5a5f6057ae85f704ab1850bef075914f68ed3aeb15a1ff1ebc0dc6a028302606092a864886f70d01090e311930173 ...
```
I've just tried and if I run this test case on 2.28, it fails.

So the bug is present in 2.28 and we do need a backport, both of this test case and of the backport.

To save a bit of time, in 2.28, the `depends_on` line should be
```
X509 CSR ASN.1 (Unsupported critical extension)
depends_on:MBEDTLS_ECDSA_C:MBEDTLS_ECP_DP_SECP256R1_ENABLED:MBEDTLS_SHA256_C
mbedtls_x509_csr_parse:"308201233081cb02010030413119301706035504030c1053656c66207369676e65642074657374310b300906035504061302444531173015060355040a0c0e41757468437274444220546573743059301306072a8648ce3d020106082a8648ce3d03010703420004c11ebb9951848a436ca2c8a73382f24bbb6c28a92e401d4889b0c361f377b92a8b0497ff2f5a5f6057ae85f704ab1850bef075914f68ed3aeb15a1ff1ebc0dc6a028302606092a864886f70d01090e311930173015060b2b0601040183890c8622020101ff0403010101300a06082a8648ce3d040302034700304402200c4108fd098525993d3fd5b113f0a1ead8750852baf55a2f8e670a22cabc0ba1022034db93a0fcb993912adcf2ea8cb4b66389af30e264d43c0daea03255e45d2ccc":"":MBEDTLS_ERR_X509_INVALID_EXTENSIONS+MBEDTLS_ERR_ASN1_UNEXPECTED_TAG
```

gilles-peskine-arm replied:
Having dug deeper, if I understand correctly, this test case has a **[requested extension](https://datatracker.ietf.org/doc/html/rfc2985#section-5.4)** which is marked as critical. Why should it be rejected? The CA doesn't need to understand requested extensions. That's the job of the verifier. All the CA needs to do is to copy the extensions into the certificate. The `critical` extension request is just a data field, not a policy for the CA.

If my understanding is correct then this test case is wrong: the expected result should be a success, since there's nothing wrong with the requested extensions.

Given the debate here we should have both a test case with a requested extension critical=true and one with critical=false, both expected to succeed. (In addition to the existing test cases with extensions without the `critical` attribute.)

With respect to backporting, we'd still want to backport the test cases to 2.28, and they would pass.

davidhorstmann-arm replied:
Hmm.
I think that the CA needs to understand the extensions to ensure that it's okay to sign the certificate, otherwise you could do the following:

1. Mallory -> CA Bob: "Please sign this CSR with this SubjectAltName extension, `dNSName=importantdomain.com`"
2. CA Bob -> Mallory: [Bob doesn't understand SubjectAltName] "Certainly, here is your certificate with your unknown extension copied in"
3. Mallory has a signed certificate with a domain they don't own, because Bob wasn't aware of the SubjectAltName extension.

This is why we currently ignore all CSR requested extensions that we do not understand, without copying them into the CSR.

mschulz-at-hilscher replied:
Hi Gilles,
thanks for your analysis. If you go back to the issue #8377, the original problem was that extensions with the optionally available critical flag were not accepted. Hence, in my initial solution (https://github.com/Mbed-TLS/mbedtls/pull/8378/files/adb3cc4d433466f00ba827d5288577e367fbf1f3#diff-dccd5ffd55ff529be4e531be08342be60a967c90fa9f1ad0aef790e615846148), I justed tried to extract the optional critical flag but then ignored it in case it was available.
In https://github.com/Mbed-TLS/mbedtls/issues/8377#issuecomment-1766389135, Tom argued that we shall not blindly accept critical extensions, in case we do not know them, hence, the callbacks were getting introduced. As no extensions are parsed in mbedTLS < 3.4.0, all extensions critical or not are accepted.

gilles-peskine-arm replied:
Thanks David, that's a very good example. It shows that we must not copy unknown extensions. But fortunately as you write, we don't copy unknown extensions: we skip them. This is true in 2.28 and in 3.5.

With this pull request, we change the behavior of unknown critical extensions to be rejected instead of ignored. With the understanding I now have, this seems reasonable to me.

All of this should be documented, though! This is a preexisting issue, but as is often the case with technical debt, we could get away with it originally, but now that it explodes in our face, we need to resolve it.

So I think what we need to do now is:

* In both 2.28 and here, update `x509_csr.h` to explain how `mbedtls_x509_csr_parse_der` handles requested extensions.
* Have test cases, in both 2.28 and here, with:
    * An unknown extension without the critical attribute.
    * An unknown extension with critical=false.
    * An unknown extension with critical=true.

mschulz-at-hilscher replied:
Hi Gilles,

I now updated the mbedTLS 3 test cases:
- An unknown extension without the critical attribute.
  - X509 CSR ASN.1 (Unsupported non-critical extension, critical undefined)
  - [Test data](http://lapo.it/asn1js/#MIIBIjCByAIBADBBMRkwFwYDVQQDDBBTZWxmIHNpZ25lZCB0ZXN0MQswCQYDVQQGEwJERTEXMBUGA1UECgwOQXV0aENydERCIFRlc3QwWTATBgcqhkjOPQIBBggqhkjOPQMBBwNCAARflLKNEzQYgzvxDEQtkTBkWdOSXnzqBuu5Igky594Rb7ZxxdLWwKN4ShKJcheu-EMucij86gqwFr22e2fO1MYSoCUwIwYJKoZIhvcNAQkOMRYwFDASBgsrBgEEAYOJDIYiAgQDAQEBMAoGCCqGSM49BAMCA0kAMEYCIQCbHosld1wYUl6WdT4e1Vh1-NYvAmxbf3DrUDetJ9yS3gIhALod_hTeavamA_djVj_QRrHNNxS1TW2vXYpyB2SX8RAU)
  - ![grafik](https://github.com/Mbed-TLS/mbedtls/assets/140500342/e94d0d09-9b93-431d-bcd2-89614d0ab6e6)
- An unknown extension with critical=false.
  - X509 CSR ASN.1 (Unsupported non-critical extension, critical=false)
  - [Test data](http://lapo.it/asn1js/#MIIBJDCBywIBADBBMRkwFwYDVQQDDBBTZWxmIHNpZ25lZCB0ZXN0MQswCQYDVQQGEwJERTEXMBUGA1UECgwOQXV0aENydERCIFRlc3QwWTATBgcqhkjOPQIBBggqhkjOPQMBBwNCAAS15xin32shkScz53xC8ma4KD5srnrfiv1WuZDBxiMuoKKkYJfCGDU7yUhESuo9AEI-hIAuKMSAmWQf6Zd837UFoCgwJgYJKoZIhvcNAQkOMRkwFzAVBgsrBgEEAYOJDIYiAgEBAAQDAQEBMAoGCCqGSM49BAMCA0gAMEUCIQDwupoIRq0KfO_Qph1fySGU3AYDekQVjeLQxWmRLAWNQwIgQh8nqfJJwWh-L6NNs_VDyFEv2SXf5a4Ahn8Tlj_9T40)
  - ![grafik](https://github.com/Mbed-TLS/mbedtls/assets/140500342/782a19e4-fdd6-4f9a-b2c6-563578c56d67)
  - Test CSR was generated using mbedTLS after editing x509_write_extension to always add the optional critical field:
  ```
    if (ext->val.p[0] != 0) {
        MBEDTLS_ASN1_CHK_ADD(len, mbedtls_asn1_write_bool(p, start, 1));
    } else {
        MBEDTLS_ASN1_CHK_ADD(len, mbedtls_asn1_write_bool(p, start, 0));
    }
  ```
- An unknown extension with critical=true.
  - X509 CSR ASN.1 (Unsupported critical extension, critical=true)
  - X509 CSR ASN.1 (Unsupported critical extension accepted by callback, critical=true)
  - X509 CSR ASN.1 (Unsupported critical extension rejected by callback, critical=true)
  - [Test data](http://lapo.it/asn1js/#MIIBIzCBywIBADBBMRkwFwYDVQQDDBBTZWxmIHNpZ25lZCB0ZXN0MQswCQYDVQQGEwJERTEXMBUGA1UECgwOQXV0aENydERCIFRlc3QwWTATBgcqhkjOPQIBBggqhkjOPQMBBwNCAATBHruZUYSKQ2yiyKczgvJLu2woqS5AHUiJsMNh83e5KosEl_8vWl9gV66F9wSrGFC-8HWRT2jtOusVof8evA3GoCgwJgYJKoZIhvcNAQkOMRkwFzAVBgsrBgEEAYOJDIYiAgEB_wQDAQEBMAoGCCqGSM49BAMCA0cAMEQCIAxBCP0JhSWZPT_VsRPwoerYdQhSuvVaL45nCiLKvAuhAiA025Og_LmTkSrc8uqMtLZjia8w4mTUPA2uoDJV5F0szA)
  - ![grafik](https://github.com/Mbed-TLS/mbedtls/assets/140500342/3c8bdcc3-f140-4e07-bdfc-5880846f1d29)

mschulz-at-hilscher replied:
In case the behavior whall be described in x509_csr.h, then the behavior or mbedtls_x509_crt_parse_der should also be described in x509_crt.h.

davidhorstmann-arm replied:
Re documenting this, I think in 2.28 the description of `mbedtls_x509_csr_parse_der()` is already correct, as [it says](https://github.com/Mbed-TLS/mbedtls/blob/mbedtls-2.28/include/mbedtls/x509_csr.h#L71):
```
* \note           CSR attributes (if any) are currently silently ignored.
```
Although perhaps it could be made more clear that this includes the requested extensions. However, the testcases would still need adding so this doesn't save us anything.

mschulz-at-hilscher replied:
In the development branch the same note exists for this function. So shall we simply remove it?
https://github.com/Mbed-TLS/mbedtls/blob/development/include/mbedtls/x509_csr.h#L78

davidhorstmann-arm replied:
In `development` we actually have changed the behaviour here so should change it to something like:
```
* \note

mschulz-at-hilscher replied:
Backport PR (https://github.com/Mbed-TLS/mbedtls/pull/8491) for the tests is ready. However, tests currently fail. This is expected as mbedtls < 3.4.0 does not care whether unsupported critical extensions exist of not when parsing a CSR.

mschulz-at-hilscher replied:
> In `development` we actually have changed the behaviour here so should change it to something like:
> 
> ```
> * \note          Any unsupported requested extensions are ignored, unless the critical flag
> *                is set, in which case the CSR is rejected.
> ```

Done.

gilles-peskine-arm commented:
@davidhorstmann-arm This is about `mbedtls_x509_csr_parse_der`. Its behavior changed in this pull request.

davidhorstmann-arm reviewed (approved):
LGTM, thanks again!

paul-elliott-arm reviewed (approved):
LGTM, I believe all concerns have been addressed.
